Updates to libGLX vendor library interface

kbrenneman opened the pull request:
This is an assortment of changes to the libGLX interface.

The `threadAttach` callback is probably the most critical, since it provides a more general workaround to a lot of broken applications.

The others are as much cleanup as anything else.

One question that's been raised is whether to keep a single version number or to split it into a major/minor pair, with the usual pattern of incrementing the minor version for backward-compatible changes.

And if we do add a minor version number, should it get passed to __glx_Main as a separate parameter (changing the function signature)? It would also be possible to just put the major and minor version numbers into the high- and low-order 16 bits.

evelikov commented:
Very good imho. Biggest 'nitpick' is that we ought to bump package version (minor/bugix for now) with each such change, thus when you guys, the amd folk, mesa or others update and rebuild things go in a controlled manner. Namely mesa configure errors out as opposed to build for 5 minutes only to fail midway.

kbrenneman commented:
A few updates:

I changed the ABI version number to be a (major, minor) pair instead of a single version number, so that it can describe any backward-compatible changes that we might make in the future.

I also bumped the package version to 0.1.0.

evelikov commented:
Wondering if one can link the whole major,minor section from the comment above__GLX_MAIN_PROTO. Also GLX_dummy should be updated to consider compatible incompatible older ABI versions - mostly for demonstration purposes.

Speaking of __GLX_MAIN_PROTO - is there any particular reason behind the macro, as opposed to a function declaration (alongside the typedef of course) ? Afaics the vendorID is not really useful so we might as well drop it, as we've bumping major ;-)

kbrenneman commented:
I don't know what the _GLX_MAIN_PROTO macro was for originally. I wouldn't object to getting rid of it, though.

Removing the vendor ID would be a bit harder. We can't use a version number to check for that, since it would change the signature of the function that takes the version number.

evelikov commented:
It feels like there is a stigma changing (m)any of the interfaces. If feels rather strange as
- in the proprietary driver, glvnd is opt only; which imho implies 'use at your own risk'
- libglvnd does not have a single release; be that alpha, or anything else

What if we (I) sort things out with distributions to not ship proprietary driver X with libgvnd Y or later. Would that be OK or there are other concerns ?

evelikov commented:
If the concern is about using the installer from the official website, there is an alternative:
- keep the _PROTO macro (greater flexibility)
- change the function name in both macros alongside the argument/typedef tweak
- probe for the compatible symbol at install time
- ...
- profit

kbrenneman commented:
Okay, I've added a few changes to update the __glx_Main function.

To start with, I removed the vendor ID and vendor name arguments, since neither one is really useful to a vendor library. I added the __GLXvendorInfo pointer as a parameter, though. If we do need to add some additional piece of vendor-specific data later on, then we can add a function to __GLXapiExports to retrieve it, without breaking compatibility.

For the second part, I rearranged things so that libGLX.so will allocate and zero the __GLXapiImports, and then the vendor library will just fill it in. That way, we can add other optional functions to the end of __GLXapiImports, again without breaking backward compatibility.

evelikov commented on src/GLX/libglxmapping.c:
```diff
@@ -510,16 +520,33 @@ __GLXvendorInfo *__glXLookupVendorByName(const char *vendorName)
             }
 
+            // Plug in the vendor imports table.
+            pEntry->imports.patchCallbacks = &pEntry->patchCallbacks;
```
Please drop this line. We appropriately set the pointer a few lines below.

evelikov commented on include/glvnd/libglxabi.h:
```diff
@@ -300,25 +300,30 @@ typedef struct __GLXapiImportsRec {
-#define __GLX_MAIN_PROTO(version, exports, vendorName)                \
-    const __GLXapiImports *__glx_Main(uint32_t version,               \
+#define __GLX_MAIN_PROTO(version, exports, vendor, imports)                \
+    __GLXapiImports *__glx_Main(uint32_t version,               \
```
Return type should be Bool, shouldn't it ?

evelikov commented on src/GLX/libglxmapping.c:
```diff
@@ -447,6 +447,37 @@ static GLboolean LookupVendorEntrypoints(__GLXvendorInfo *vendor)
+            || pEntry->imports.getProcAddress == NULL
+            || pEntry->imports.getDispatchAddress == NULL
+            || pEntry->imports.setDispatchIndex == NULL)
+    {
```
Nitpick: I'd move the opening curly bracket to the previous line - ... == NULL) {
It seems to be the style used throughout the project.

evelikov commented on include/glvnd/libglxabi.h:
```diff
@@ -300,25 +300,30 @@ typedef struct __GLXapiImportsRec {
+     * Like the __GLXapiImports struct itself, this struct is allocated and
+     * zeroed by libGLX. The vendor library should assign function pointers
+     * to the members of this struct, but should not change the
+     * \c patchCallbacks pointer itself.
```
Make it a constant pointer (as opposed to the original constant data), so that users get a lovely warning ;-) Otherwise this comment might go unnoticed ?

evelikov commented on src/GLX/libglxmapping.c:
```diff
@@ -447,6 +447,37 @@ static GLboolean LookupVendorEntrypoints(__GLXvendorInfo *vendor)
+
+    // Check to see whether this vendor library can support entrypoint
+    // patching.
+    if (pEntry->imports.patchCallbacks != NULL) {
```
I'm confused. Barring vendors overwriting the (const) pointer `patchCallbacks`, how can this be null ?

Fwiw I'm inclined to keep the following hunk in this function. It makes things more obvious

`pEntry->imports.patchCallbacks = &pEntry->patchCallbacks`

evelikov commented on tests/GLX_dummy/GLX_dummy.c:
```diff
@@ -664,37 +664,27 @@ static GLboolean dummyInitiatePatch(int type,
-    .getProcAddress = dummyGetProcAddress,
-    .getDispatchAddress = dummyGetDispatchAddress,
-    .setDispatchIndex = dummySetDispatchIndex,
+    if (version <= GLX_VENDOR_ABI_VERSION) {
```
Not sure if I mentioned it previously - we ought to update this to honour the new major/minor version. Otherwise some unsuspecting vendor will do copy it and things will explode down the line.

evelikov commented:
The series look very good. There are a few small suggestions (some of which obsoleted by the next patches in the series) all of which could be done on top, if needed.

An alternative to the last commit (removing the __GLdispatchPatchCallbacks pointer), is to embed __GLdispatchPatchCallbacks into __GLXapiImports. It will give nice separation/clarity, although it's quite subjective suggestion.

kbrenneman commented:
I thought about embedding the __GLdispatchPatchCallbacks struct, but that would lead back to the original problem of breaking compatibility: Adding functions to the end of __GLdispatchPatchCallbacks would be equivalent to adding functions to the middle of __GLXapiImports.

evelikov commented:
Indeed - extending __GLXapiImports and then __GLdispatchPatchCallbacks will lead to ABI chaos. I assumed that the latter is baked on stone (will never change) - silly me.

Can you please hint/elaborate on the (ABI) downsides of using a `__GLdispatchPatchCallbacks const *patchCallbacks;` - i.e. without "GLX: Add the patch callbacks directly to __GLXapiImports." applied ? The only issue that comes to mind - using an incorrect size during memcpy() - is a programmer error. Mostly curious - I don't feel strongly attached to `*patchCallbacks`.

kbrenneman commented:
Mainly just that a single flat struct is simpler and has fewer things to mess up.

kbrenneman commented:
Now that I think about it, if __GLXapiImports still had the patchCallbacks pointer, then a vendor library couldn't use memcpy at all to fill it in, since that would overwrite the patchCallbacks pointer. Calling memcpy for the first part of it might work, but that would be more than a little awkward, and gets even more so if additional pointers are added to the end of __GLXapiImports.

With a flat struct, the vendor library could define a static __GLXapiImports table and just blindly memcpy the whole thing.

kbrenneman commented:
One more commit to fix the __GLX_MAIN_PROTO macro and to make GLX_dummy.c closer to what a real vendor library would do.
